prime-router/src/main/kotlin/FileNameTemplate.kt

```diff
@@ -0,0 +1,181 @@
+            val dt = OffsetDateTime.now()
+            formatter.format(dt)
+        } catch (_: Exception) {
+            ""
+        }
+    }
+}
+
+open class FileNameTemplate(
+    val elements: List<String>,
+    val lowerCase: Boolean? = null,
+    val upperCase: Boolean? = null,
+    val name: String? = null
+) {
+    fun getFileName(
+        translatorConfig: TranslatorConfiguration? = null
+    ): String {
+        val fileName = StringBuilder()
+        val parsedElements = fixupFileNameElements(elements)
+        parsedElements.forEach {
+            when (it.first) {
+                is FileNameElement -> {
+                    val e = it.first as FileNameElement
+                    fileName.append(e.getElementValue(it.second, translatorConfig))
+                }
+                is String -> {
+                    fileName.append(it.first)
+                }
+                else -> {
+                }
+            }
+        }
+        return when {
+            (lowerCase == true) -> fileName.toString().lowercase()
+            (upperCase == true) -> fileName.toString().uppercase()
+            else -> fileName.toString()
+        }
+    }
+
+    companion object {
+        // this is the collection of file name elements that can be invoked
+        private val fileNameElements: List<FileNameElement> = listOf(
+            Literal(),
+            ReceivingOrganization(),
+            CreatedDate(),
+            RegexReplace(),
+            FileUuid(),
+            SchemaBaseName(),
+            ProcessingModeCode(),
+        )
+
+        private fun findFileNameElement(elementName: String): FileNameElement? {
+            return fileNameElements.firstOrNull {
+                it.name == elementName
+            }
+        }
+
+        fun parseFileNameElement(fileNameElement: String): Pair<String, List<String>> {
+            // Using a permissive match in the (arg1, arg2) section, to allow most regexes to be passed as args.
+            // Somehow this works with "(?i).*Pima.*", I guess because the \\x29 matches rightmost ')' char
+            // that said, this allows for the passing in of bare strings, so if the match fails, then we
+            // just pass through a pair of the failing value to an empty list
+            val match = Regex("([a-zA-Z0-9]+)\\x28(.*)\\x29").find(fileNameElement)
+                ?: return fileNameElement to emptyList()
+            return match.groupValues[1] to match.groupValues[2].split(',').map { it.trim() }
+        }
+
+        fun fixupFileNameElements(elements: List<String>): List<Pair<Any, List<String>>> {
+            // given a list of elements in the file name template, try to cast each
+            // to a function, or to just the literal string
+            return elements.map {
+                val parsedElement = parseFileNameElement(it)
+                Pair(findFileNameElement(parsedElement.first) ?: it, parsedElement.second)
+            }
+        }
+    }
+}
```

Comment:
`FileNameElement`s act just like mappers and filters and expand on our metalanguage we're building here in ReportStream. The really critical piece here is that a file name template can contain both callable functions AND bare strings. `parseFileNameElement` attempts to parse each item in the yaml array as a file name element, and if it doesn't match the regex it will assume it's a string. There is also a `literal` `FileNameElement` that lets you force something to be a literal string if you want to be explicit about it.

Reply:
And what if someone wants "()" in their filename?!   ;)